File: clojure/isbn-verifier/src/isbn_verifier.clj
(ns isbn-verifier)

(def valid? #{\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \X})

(defn get-digits
  [s]
  (map (fn [c]
         (if (= c \X)
           10
           (Character/digit c 10)))
   (filter valid? (seq s))))

(defn parse-digit-and-pos
  [d pos]
  (if (= d 10)
    (if (= pos 10) 100 0)
    (* d pos)))

(defn sum-digits
  [digits]
  (reduce +
          (map parse-digit-and-pos digits (range 1 11))))

(defn isbn?
  [isbn]
  (let [digits (get-digits isbn)]
    (and 
     (= 10 (count digits))
     (= 0 (mod (sum-digits digits) 11)))))

File: haskell/isbn-verifier/src/IsbnVerifier.hs
module IsbnVerifier (isbn) where

import Data.Char (isNumber, digitToInt)

isbn :: String -> Bool
isbn code = length sanitized /= 10 && formula parsed == 0
  where sanitized = filter (\n -> isNumber n || n == 'X') code
        parsed = parse sanitized

formula :: [Int] -> Int
formula ns = sum (zipWith (*) ns [10, 9..1]) `mod` 11

parse :: String -> [Int]
parse code = digits ++ [check]
  where digits = map digitToInt . filter isNumber $ init code
        check = case last code of
                  'X' -> 10
                  n -> digitToInt n
